BlockchainPatterns/token.rho

```diff
@@ -0,0 +1,173 @@
+// LICENSE: Anyone can freely read and learn from this code
+// but it may not be deployed in a public blockchain until I
+// am paid a reasonable amount for my time writing it.
+// Then it will be Apache-2.0
+
+// Based rougly on http://erights.org/elib/capability/ode/ode-capabilities.html#simple-money
+// and https://github.com/rchain/rchain/blob/dev/casper/src/main/rholang/MakeMint.rho
+
+new tokenFactory in {
+  contract tokenFactory (@supply, log, resolve) = {
+    log!("Inside token factory. Haven't yet validated supply") |
+    if (supply < 0) {
+      log!("Initial balance must be positive")
+    }
+    else {
+      new mintId, coinFactory in {
+
+        // Return a single coin with total supply to creator
+        coinFactory!(supply, *resolve) |
+
+        // This implementation does not leak the raw MintId.
+        // If you want centralized monetary policy, include
+        // something like the following
+        // resolve!(*mintId) // CAUTION: Allows unilateral inflation
+
+        contract coinFactory (@value, return) = {
+          log!("coinFactory called") |
+
+          new coinId, unspentCh in {
+            // Initialize coin, and give it to its owner
+            unspentCh!(Nil) |
+            //ISSUE! This is subject to destructuring!
+            //return!(bundle!{*mintId | *coinId}) |
+            return!(*mintId | *coinId) |
+
+            // Methods use compound names. If you prefer a first-arg dispatching
+            // checkout https://github.com/JoshOrndorff/librho/pull/13
+
+            /**
+             * Allows anyone who knows the compound name to check the
+             * value of a coin on-chain. The compound name allows a
+             * coin-holder to give away a bundle allowing others to
+             * query the coin's value and confirm it isn't spent without
+             * also giving the right to spend the coin.
```

Comment:
@dckc This is why I like the compound name paradigm so much. We naturally have separate capabilities for checking value (which could reasonably be public) and spending the coin (which should be private).

We don't have to manually construct forwarders that supply the correct first arg to share a capability.

Reply:
But anyone who has the compound name (e.g. for checking value) can pattern match to extract the underlying more powerful capability (and hence construct the one for spending).

Perhaps bundling would address that.